model/DataObjectSchema.php

```diff
@@ -0,0 +1,377 @@
+<?php
+
+use SilverStripe\Framework\Core\Configurable;
+use SilverStripe\Framework\Core\Injectable;
+use SilverStripe\Model\FieldType\DBComposite;
+
+/**
+ * Provides dataobject and database schema mapping functionality
+ */
+class DataObjectSchema {
+	use Injectable;
+	use Configurable;
+
+	/**
+	 * Default separate for table namespaces. Can be set to any string for
+	 * databases that do not support some characters.
+	 *
+	 * Defaults to \ to to conform to 3.x convention.
+	 *
+	 * @config
+	 * @var string
+	 */
+	private static $table_namespace_separator = '\\';
+
+	/**
+	 * Cache of database fields
+	 *
+	 * @var array
+	 */
+	protected $databaseFields = [];
+
+	/**
+	 * Cache of composite database field
+	 *
+	 * @var array
+	 */
+	protected $compositeFields = [];
+
+	/**
+	 * Cache of table names
+	 *
+	 * @var array
+	 */
+	protected $tableNames = [];
+
+	/**
+	 * Clear cached table names
+	 */
+	public function reset() {
+		$this->tableNames = [];
+		$this->databaseFields = [];
+		$this->compositeFields = [];
+	}
+
+	/**
+	 * Get all table names
+	 *
+	 * @return array
+	 */
+	public function getTableNames() {
+		$this->cacheTableNames();
+		return $this->tableNames;
+	}
+
+	/**
+	 * Given a column name, build the "Table"."Column" string appropriate for querying with.
+	 * This method will correctly match the column to the data class and table.
+	 *
+	 * Note that any fixed_fields will use the base table. Use quotedID() for joining
+	 * on the ID of non-base tables.
+	 *
+	 * @param string $class Class name
+	 * @param string $column Column name
+	 * @return string The quoted column, if this column belongs to this model
+	 */
+	public function quotedColumn($class, $column) {
+		$table = $this->tableForField($class, $column);
+		if($table) {
+			return "\"{$table}\".\"{$column}\"";
+		}
+		return null;
+	}
+
+	/**
+	 * Quotes the ID column of the specific table.
+	 * Unlike quotedColumn, does not force "ID" to the base table.
```

Comment:
I feel that this should be given a name to discourage its use except for in building up the multi-table inheritance joins. E.g. `quotedSubclassID()`?

Regular use should always refer to the base table's ID, so that a missing record in a subtable doesn't break things.